### ISSUE TITLE
Determine the GalSim atmospheric input parameters for the Cerro Pachon site

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/imSim#92]
@jmeyers314 has made good progress on #8 and we expect to have a ray-traced atmospheric mode in GalSim soon.

The GalSim atmospheric simulation will need inputs appropriate for our site.  Josh pointed out we will need things like number of layers, layer weights, altitudes, velocities, the net seeing, etc etc.

Has anyone thought about where to source those? @bxin and @cclaver and @SimonKrughoff do you have any useful input to the discussion?

Related to this: I think we want to be able to drive the overall seeing by the OpSim visit variable.  So, we need to think about how to go from that input to the model parameters that would give us that mean seeing.

**Comment by SimonKrughoff** [comment]
I'm not sure I have anything concrete, but one thing that came up at the sims review a few years back is that having any more layers than three is probably not supportable by current observations.  That may not be the case anymore.

**Comment by jmeyers314** [comment]
Here's my untested probably buggy python/pseudo-code proposal for this:

```python
def getAtm(seeing):
    """
    @param seeing  Desired seeing.
    @returns  galsim.PhaseScreenList 
    """
    # Use Ellerbroek 2002, JOSA Vol 19 No 9  Gemini site testing data
    altitudes = [0.0, 2.58, 5.16, 7.73, 12.89, 15.46]  # km, from Ellerbroek
    weights = [0.652, 0.172, 0.055, 0.025, 0.074, 0.022]  # from Ellerbroek
    # Add 10% random variability to weights (this is admittedly pretty arbitrary).
    weights = [np.abs(w*np.random.normal(1, 0.1)) for w in weights]
    # Renormalize
    wsum = np.sum(weights)
    weights = [w/wsum for w in weights]
    
    # Draw a single common outer scale for all layers from a log normal
    L0 = np.random.lognormal(mean=25, sigma=10)
    # Restrict between 10m and 100m 
    L0 = np.clip(L0, 10, 100)
    
    # uniformly random speeds and directions
    max_speed = 20   # m/s
    speeds = [np.random.uniform() * max_speed for _ in range(6)]
    directions = [np.random.uniform() * 360 for _ in range(6)]  # degrees

    # Solve Martinez 2010 A&A 516, A90 equation (4) for r0 given wavelength, L0, and desired seeing
    # seeing = 0.976 wavelength/r0 * sqrt(1-2.183(r0/L0)**0.356)
    r0 = solveMartinezEq4(wavelength, L0, seeing)

    atm = galsim.Atmosphere(
            altitude=altitude, r0_weights=weights, r0=r0, L0=L0, 
            speeds=speeds, directions=directions)
    return atm
```

**Comment by cwwalter** [comment]
Great.  So this code would live in imSim and then pass that atmosphere to GalSim is that correct?

If so a few things to consider:

- Are the weights/heights and # of entries basically a fixed thing do you think? In no, do you think we should be able to configure them for the run in the config file?

- Is it worth writing the values of the randomly chosen outer scale, speeds, and directions into the header keywords? This way would could get this truth into the tables for later analysis. In general I am a believer of getting all of the truth information we can.  But, it has to be worthwhile enough that we might look at it later.

**Comment by jmeyers314** [comment]
>Great. So this code would live in imSim and then pass that atmosphere to GalSim is that correct?

Yep, I think think that's right.

>Are the weights/heights and # of entries basically a fixed thing do you think? In no, do you think we should be able to configure them for the run in the config file?

I don't have a more intelligent way of picking weights/heights other than to pull these from Ellerbroek (this is also what Jee&Tyson 2011 used, btw).  That said, if we do come up with something more intelligent in the future, then placing these in a config file would make the current code forwards-compatible, which sounds good to me.

>Is it worth writing the values of the randomly chosen outer scale, speeds, and directions into the header keywords? This way would could get this truth into the tables for later analysis. In general I am a believer of getting all of the truth information we can. But, it has to be worthwhile enough that we might look at it later.

I think these are worth recording.  It would be interesting to check if correlated PSF patterns are aligned with the dominant layer wind direction, for example.  ~30 floats isn't too painful, but it might be nice to take advantage of the fact that these are constant per exposure and not repeat them into all 189 CCD image headers.  Is there a natural once-per-exposure output somewhere already?

**Comment by cwwalter** [comment]
> think these are worth recording. It would be interesting to check if correlated PSF patterns are aligned with the dominant layer wind direction, for example. ~30 floats isn't too painful, but it might be nice to take advantage of the fact that these are constant per exposure and not repeat them into all 189 CCD image headers. Is there a natural once-per-exposure output somewhere already?

Ah right... Basically no.  We should probably do this but exactly how depends a bit on how things are setup to run at scale (as opposed to on your laptop) @jchiang87 what exactly happens in the pipeline when a exposure is run?  Is each sensor it's own job? Or do they get dispatched by one process somehow to different cores somehow?

**Comment by jchiang87** [comment]
Currently the plan is to run each sensor visit as a separate job, but the instance catalog generation would be on a per-visit basis over the whole focal plane, so we could certainly write the atmosphere parameters for a given visit to a file (or even directly to a db table) for later retrieval.

**Comment by cwwalter** [comment]
So that code snippet would need to be run once to choose the random values. It can't run each time in separate jobs for each sensor (unless we pass a seed..) 

Maybe it would need to be moved up into instance catalog generation and specify the parameters there?

**Comment by jchiang87** [comment]
Right, this is what I  have just proposed:   For each visit, it would be run exactly once to generate the instance catalog that all of the sensors would use.  I'm not sure we want to put it into the instance catalog generation code itself (i.e. in sims_GCRCatSimInterface), because that would induce unnecessary coupling between that code and galsim.  If this code needs to adjust or add parameters to the instance catalog header that imsim reads, I think it would be better to run it as a separate step after the initial instance catalog was generated.

**Comment by cwwalter** [comment]
OK that sounds good... need to go to the airport now.  We would also want this work running in a mode on your laptop where the sensors are handled in one job...

**Comment by jmeyers314** [comment]
>Currently the plan is to run each sensor visit as a separate job.

I think this is fine, but want to verify that we can still multiprocess within this job.  To take advantage of the 68 cores per 96 GB of memory on KNL (and use the atmospheric PSF) we need to have at least a few cores per job that all share the same atmospheric screens in memory.  I think 17 cores per job would be a natural choice since it evenly divides 68.

**Comment by jchiang87** [comment]
Actually, when I said one job per sensor visit, I meant one process per sensor visit in the context of using the multiprocessing module on KNL.  We should discuss this week how to get the cores to share the atmosphere screens in memory in this case.

**Comment by jmeyers314** [comment]
Okay.  I suppose 1 is also a perfectly good divisor of 68..., though I'd worry a bit more about cores becoming idle in this case.

**Comment by jchiang87** [comment]
That's not what I had in mind.  I've been packing the nodes as much as possible with imsim jobs to fill the memory usage, i.e., using as many cores as possible per node.

**Comment by jmeyers314** [comment]
Sorry, I meant 1 core per sensor-visit is okay; not one core per job.  We definitely want to pack the nodes.  For the PSF validation stuff I've been able to occupy all 68 on a node.

**Comment by cwwalter** [comment]
@jmeyers314 Put in his best guess for values and an initial setup is now being used.  This is closed by #101.
